File: typical90/typical90_bc.py
def main():
    ans = 0
    stack = [(A[i], i, 1) for i in range(N-4)]
    while stack:
        mod, idx, cnt = stack.pop()
        if cnt == 5:
            ans += (mod == Q)
            continue
        for j in range(idx+1, N-4 + cnt):
            nxt = (mod * A[j]) % P
            stack.append((nxt, j, cnt+1))
    return print(ans)


if __name__ == '__main__':
    N, P, Q = map(int, input().split())
    A = list(map(lambda x: int(x) % P, input().split()))

    main()

File: typical90/typical90_bc.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using tlll = tuple<ll, ll, ll>;

#define rep(i, n) for (ll i = 0; i < ll(n); i++)
#define repr(i, a, b) for (ll i = ll(a); i < ll(b); i++)

ll dfs(const vector<ll>& vec, ll p, ll q) {
    ll n = vec.size();
    vector<tlll> st(0);
    rep(i, n - 4) st.push_back(tlll(vec[i], i, 1));

    ll res = 0;
    ll mod, idx, cnt;
    while (!st.empty()) {
        tie(mod, idx, cnt) = st.back();
        st.pop_back();
        if (cnt == 5) {
            if (mod == q) res++;
            continue;
        }
        repr(j, idx + 1, n - 4 + cnt) {
            ll nxt = (mod * vec[j]) % p;
            st.push_back(tlll(nxt, j, cnt + 1));
        }
    }
    return res;
}

int main() {
    ll n, p, q;
    cin >> n >> p >> q;
    vector<ll> a(n);
    rep(i, n) cin >> a[i];

    ll ans = dfs(a, p, q);
    cout << ans << endl;
    return 0;
}
